Feature: add forgit show command

carlfriedrich opened the pull request:
Use the new command in git log enter. This fixes the display of diffs for merge commits.

Fixes #416.

<!-- NOTE: forgit.plugin.zsh & forgit.plugin.sh share the same code. You should make sure the changes work in both `zsh` & `bash` -->

<!-- Check all that apply [x] -->

## Check list

- [x] I have performed a self-review of my code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

## Description

<!-- Please include a summary of the change (and the related issue, if any). Please also include relevant motivation and context when necessary. -->

## Type of change

- [x] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Documentation change

## Test environment

- Shell
    - [x] bash
    - [ ] zsh
    - [ ] fish
- OS
    - [x] Linux
    - [ ] Mac OS X
    - [ ] Windows
    - [ ] Others:

carlfriedrich commented:
Not sure about the fish completions, we don't have them for `forgit diff` either. @cjappl or @sandr01d do you know more about this?

cjappl commented:
> Not sure about the fish completions, we don't have them for `forgit diff` either. @cjappl or @sandr01d do you know more about this?

Hmm, good find. Would you file a ticket for this and assign it to me and I can fix it when I get time? I can confirm git diff does not behave with completions on fish.

wfxr reviewed (approved):
LGTM.

sandr01d reviewed (changes requested):
Looking forward to see this feature merged, I definitely need this :smiley:
In addition to things I've pointed out in the code, there are two things I've noticed:

1. The show command has not been added to _conf.d/forgit.plugin.fish_ and as a result does not work in fish. Might the issue you were having with fish completions be related to this? 
2. Completions for zsh are missing. If I remember correctly, you're not a zsh user @carlfriedrich, so if you don't want to go through the hassle of setting up zsh on your system I'd offer to implement a zsh completion and add it to this PR. Just let me know.

carlfriedrich commented on bin/git-forgit:
```diff
@@ -339,6 +339,65 @@ _forgit_diff() {
+        $FORGIT_FZF_DEFAULT_OPTS
+        +m -0 --bind=\"enter:execute($FORGIT show_enter {} $escaped_commit | $FORGIT pager enter)\"
+        --preview=\"$FORGIT show_view {} '$_forgit_preview_context' $escaped_commit\"
+        --bind=\"alt-e:execute-silent($FORGIT edit_diffed_file {})+refresh-preview\"
```
Good catch, thank you! I updated this accordingly.

carlfriedrich commented on README.md:
```diff
@@ -101,6 +101,8 @@ Then add the following to your shell's config file:
 
 - **Interactive `git diff` viewer** (`gd`)
 
+- **Interactive `git show` viewer** (`gs`)
```
Hm, I see. Not sure about this, though. Is ghostscript actually commonly installed? I don't have it installed, and it does not ship with Ubuntu by default for example. And furthermore, if you have it installed, do you actually use it by calling the command interactively, or isn't it more a kind of util used by scripts?
IMO we could stick with `gs`, since I don't particularly like the alternatives (and would override it to `gs` anyway). Users can still override it to something different in their shell config if it interferes with ghostscript.
However, since I am not using the shell aliases myself anyway, I am open to any other suggestion.

sandr01d replied:
> Is ghostscript actually commonly installed?

I think it is, at least as a dependency. On my system it is a dependency of my printers drivers and LaTeX.

> And furthermore, if you have it installed, do you actually use it by calling the command interactively, or isn't it more a kind of util used by scripts?

That is true though, I never use it interactively.

Thinking more about this, it came to my mind that a lot of people probably use `gs` as an alias for `git status`. At least that's what I've seen in the past.  So I would still favor using a different alias to not break peoples workflows.

carlfriedrich replied:
`gs` being a common alias for `git status` is a good point. I went with `gso` (for "git show object") now. Deal?

sandr01d replied:
Works for me, thanks!

carlfriedrich commented:
Thanks for your reviews everyone, especially for the good finds @sandr01d.

>     1. The show command has not been added to _conf.d/forgit.plugin.fish_ and as a result does not work in fish. Might the issue you were having with fish completions be related to this?

I am sorry, I simply forgot this. Updated it in the new commit version. However, I did not test the fish completions, I just did not add anything in the fish completions file. I was looking for something to copy from `forgit diff`, but then noticed that that's missing as well.

>     2. Completions for zsh are missing. If I remember correctly, you're not a zsh user @carlfriedrich, so if you don't want to go through the hassle of setting up zsh on your system I'd offer to implement a zsh completion and add it to this PR. Just let me know.

Good point, I added them in the new commit version, but haven't tested them. Can you give them a quick test?

> Hmm, good find. Would you file a ticket for this and assign it to me and I can fix it when I get time? I can confirm git diff does not behave with completions on fish.

@cjappl Yep, thanks a lot, I will do so.

sandr01d commented:
Thanks for the updates @carlfriedrich. Fish is working now (except completions, but we can handle this in #418, so no objections from my side). Zsh completions are working fine too.
There is one additional thing I've noticed while testing: the current implementation does not work well with tags. The reason for this is that git provides some additional information when showing a tag. E.g. `git show --pretty="" --name-status 24.10.0` produces the following output:

```
tag 24.10.0
Tagger: github-actions <github-actions@users.noreply.github.com>

Release 24.10.0
M	README.md
M	bin/git-forgit
M	completions/_git-forgit
M	completions/git-forgit.bash
M	completions/git-forgit.fish
M	conf.d/forgit.plugin.fish
M	forgit.plugin.zsh
```

While this probably is an edge case, we should be able to filter out the additional information quite easily.

carlfriedrich commented:
> Thanks for the updates @carlfriedrich. Fish is working now (except completions, but we can handle this in #418, so no objections from my side). Zsh completions are working fine too. There is one additional thing I've noticed while testing: the current implementation does not work well with tags. The reason for this is that git provides some additional information when showing a tag. E.g. `git show --pretty="" --name-status 24.10.0` produces the following output:

@sandr01d Good catch, thank you! I updated the call producing the file list to use the commit object instead of the tag. I kept the original call for the preview and enter commands, though. IMO if you call `git show` on a tag you actually might want to see the tag information, so now it is displayed above the diff for each file. Are you fine with that?

sandr01d commented:
>  I kept the original call for the preview and enter commands, though. IMO if you call git show on a tag you actually might want to see the tag information, so now it is displayed above the diff for each file. Are you fine with that?

Yes, absolutely. Having the tag information displayed is a big plus IMO.

> Thinking more about this I am asking myself whether people might be interested in the commit message as well when calling forgit show. Should we display it somehow? Maybe implement a preview toggle so that we can switch between the commit message and the file diff (using [transform action](https://github.com/junegunn/fzf/releases/tag/0.45.0))?

I think this is a great idea! Wasn't aware of the `transform` action — what a neat feature. We'd have to check for the installed fzf version tough. I suspect a lot of users running versions older than 0.45.0.

carlfriedrich commented:
@sandr01d So I gave it a shot and added a keybinding to toggle between the commit message and the commit diff. Works pretty well IMO and I already see myself using it a lot.

I also added a border to the preview window indicating what is currently shown. And furthermore I added a border displaying the keybindings.

I think maybe the latter might trigger a more general discussion (as I already suggested someting similar in #259) and maybe we might split that part out to a new MR.

I did not yet check the fzf version, yet (we need at least 0.49.0 now). I wanted to discuss how we continue on this, first. Maybe we should merge the show feature as it was before the new commit (because it already worked before) and add everything else in one or more follow-ups.

Let me know what you think and how you like the proposal.

sandr01d commented on bin/git-forgit:
```diff
@@ -339,6 +339,77 @@ _forgit_diff() {
+        +m -0 --bind=\"enter:execute($FORGIT show_enter {} $escaped_commit | $FORGIT pager enter)\"
+        --preview=\"$FORGIT show_view {} '$_forgit_preview_context' $escaped_commit\"
+        --bind=\"alt-e:execute($FORGIT edit_diffed_file {})+refresh-preview\"
+        --bind 'ctrl-t:transform:[[ ! \$FZF_PREVIEW_LABEL =~ Diff ]] &&
+                echo \"change-preview-label( Diff )+refresh-preview\" ||
+                echo \"change-preview-label( Commit Message )+refresh-preview\"'
+        --preview '[[ \$FZF_PREVIEW_LABEL =~ Diff ]] &&
+                $FORGIT show_view {} '$_forgit_preview_context' $escaped_commit ||
+                git show --quiet --color=always \${FZF_PROMPT%% *}'
```
I don't like that this introduces more deferred execution again. In this case we'd actually also have to do weird escaped quoting:

```suggestion
        --bind 'ctrl-t:transform:[[ ! \"\$FZF_PREVIEW_LABEL\" =~ Diff ]] &&
                echo \"change-preview-label( Diff )+refresh-preview\" ||
                echo \"change-preview-label( Commit Message )+refresh-preview\"'
        --preview '[[ \"\$FZF_PREVIEW_LABEL\" =~ Diff ]] &&
                $FORGIT show_view {} '$_forgit_preview_context' $escaped_commit ||
                git show --quiet --color=always \"\${FZF_PROMPT%% *}\"'
```

How about we move the part for the preview into `_forgit_show_view` or create a separate function for it? I think this would make it easier to maintain.

carlfriedrich replied:
Moving the preview part to a new function was a very good idea, thanks for that! Implemented that in the new version. I left the `--bind` argument as deferred code, though, as this is exactly how `fzf` shows it in their examples and the echoed strings would not make sense without the context of `--bind`. Do you agree?

sandr01d replied:
Yes, I think there is now way to move the `--bind` arguments out of deferred code, which is fine for me. We should probably double quote `$FZF_PREVIEW_LABEL` in the deferred code.

carlfriedrich replied:
I added the double quotes like you suggested. Looks a bit ugly, but works. I wanted to keep the double quotes for the complete `--bind` argument to be consistent with all the other args, so I had to double escape the quotes now.

sandr01d reviewed (changes requested):
I find this super useful, well done @carlfriedrich! See my comments below regarding implementation details.
I think moving the second commit to a new PR would probably be best. E.g., I'm not a big fan of the bottom border and would prefer to have the key bindings displayed without it - but don't know if fzf supports this. With a new PR, we could also implement showing the existing keybindings of other functions (editing files with diff and add).

sandr01d commented:
> I also changed the keybinding from ctrl-t to alt-t, because we also use alt-e for edit. Ok for you?

Yeah, makes sense to keep the keybinding consistent with what we already have.

Once the opened thread above is resolved the only thing missing would be checking for the fzf version, so we don't break the function for users of older fzf versions.

carlfriedrich commented:
@sandr01d Thanks again. Please check if the quotes are fine for your and resolve the thread in case.

Concerning the fzf version check: do you have an opinion what we do when the version is below 0.49.0? Should we exit forgit with an error immediately or should we just disable the features we cannot use with older versions?
I would prefer the former, as partially disabling features would bloat the code, and basically everybody should be able to install a recent fzf version. There might be lazy people relying on old versions in their Ubuntu's apt repo ([Ubuntu 24.04 packages fzf 0.46.1](https://launchpad.net/ubuntu/oracular/+source/fzf)), but is that a reason not to move on with forgit? I mean, these people still can install old versions of forgit working with their fzf version as well.
WDYT?

Also pinging @cjappl and @wfxr for an opinion.

sandr01d commented:
> Concerning the fzf version check: do you have an opinion what we do when the version is below 0.49.0? Should we exit forgit with an error immediately or should we just disable the features we cannot use with older versions?
I would prefer the former, as partially disabling features would bloat the code, and basically everybody should be able to install a recent fzf version. There might be lazy people relying on old versions in their Ubuntu's apt repo ([Ubuntu 24.04 packages fzf 0.46.1](https://launchpad.net/ubuntu/oracular/+source/fzf)), but is that a reason not to move on with forgit? I mean, these people still can install old versions of forgit working with their fzf version as well.
WDYT?

This is a tough one. I agree that introducing version checks for specific features of fzf would make maintenance difficult if we decide to implement similar features like the one in question for other functions (which is something I'd like to see). My only concern is that there might be some users in restricted environments that are not able to install a different version of fzf than their distributions package manager provides. We had a similar case with #328. I wonder how many users would be negatively affected by such a change. Maybe it would be worth to open a discussion and start a poll on which distributions/versions people use? Not sure whether we'd see many responses, but might be worth a try.
I am currently leaning more towards requiring a newer version of fzf for forgit in general, because I think we should not hold back development. Would be interested to know what the other maintainers think.

cjappl commented:
Sorry for the delay! Thanks for pinging :) 

> Concerning the fzf version check: do you have an opinion what we do when the version is below 0.49.0? Should we exit forgit with an error immediately or should we just disable the features we cannot use with older versions?
I would prefer the former, as partially disabling features would bloat the code, and basically everybody should be able to install a recent fzf version. There might be lazy people relying on old versions in their Ubuntu's apt repo ([Ubuntu 24.04 packages fzf 0.46.1](https://launchpad.net/ubuntu/oracular/+source/fzf)), but is that a reason not to move on with forgit? I mean, these people still can install old versions of forgit working with their fzf version as well.
WDYT?

Generally in agreement with everything said here. I think we should do a (possibly major) version bump of forgit, and take in the new fzf.

> Should we exit forgit with an error immediately

This is my preference. I am against bloating the code with fzf version check

> I mean, these people still can install old versions of forgit working with their fzf version as well.

This is the key I think. This is why we version our code, so people can fall back to an older one. **Do we have to do a major version number bump as this is a backwards incompatible change?  (That would be my first thought)**

I think one of the possible good things here is that forgit is not really used as a CI/CD tool, but on devs machines directly where they have more control. I think this probably means we can be more aggressive moving forward than something crucial to many people's build pipelines.

If we get many many reports we can always roll back and remove the feature, and I will eat my words :)

carlfriedrich commented:
@sandr01d @cjappl Thanks for your opinions!

> My concern is that there might be some users in restricted environments that are not able to install a different version of fzf than their distributions package manager provides.

I don't see that, actually. Linux is flexible and `fzf` is distributed as a single binary. Everybody should be able to put it in their home directory under something like `~/.local/bin` and add that to their `PATH` (that's how I have it installed, it's actually part of my [dotfiles](https://github.com/carlfriedrich/dotfiles/tree/main/local/bin)).
The most user-friendly way is possibly to use `fzf`'s own `install` script, which is documented in their [README](https://github.com/junegunn/fzf?tab=readme-ov-file#using-git). We could point to that in our `README` as well.

> We had a similar case with https://github.com/wfxr/forgit/issues/328. I wonder how many users would be negatively affected by such a change.

This was actually about `git`, which is a more common tool to have installed via the package manager than `fzf` I would argue.

> Do we have to do a major version number bump as this is a backwards incompatible change? (That would be my first thought)

Since we don't use semantic versioning, there is no such thing as major versions in `forgit`. I think we should highlight this in the release notes and add it to the `README`, and we should make sure that the new version dependency is correctly modeled in the AUR and brew recipes.

If that's fine with you then, I would implement the hard version check.

sandr01d commented:
Hey @carlfriedrich, sorry for the delay.

> If that's fine with you then, I would implement the hard version check.

Yes, please go ahead.

carlfriedrich commented:
@sandr01d @cjappl I have implemented the version check and added it to the README as well as a separate commit, which will appear in the release notes. Can you check if that works for you? Any suggestions for improvements?

cjappl reviewed (approved):
LGTM - good work

sandr01d commented on bin/git-forgit:
```diff
@@ -12,6 +12,21 @@
+    exit 1
+fi
+
+# Check fzf version
+required_fzf_version="0.49.0"
+higher_fzf_version=$(printf '%s\n' "$required_fzf_version" "$installed_fzf_version" | sort -V | head -n1)
+if [[ "$higher_fzf_version" != "$required_fzf_version" ]]; then
+    echo "fzf version $required_fzf_version or higher is required. You have $installed_fzf_version."
+    exit 1
+fi
```
This works fine, but I think what we store in `$higher_fzf_version` is actually the lower version. `sort -V` sorts lowest →highest and we store the first line. The check is fine, but I think we should rename the variable.

```suggestion
# Check fzf version
required_fzf_version="0.49.0"
lower_fzf_version=$(printf '%s\n' "$required_fzf_version" "$installed_fzf_version" | sort -V | head -n1)
if [[ "$lower_fzf_version" != "$required_fzf_version" ]]; then
    echo "fzf version $required_fzf_version or higher is required. You have $installed_fzf_version."
    exit 1
fi
```

carlfriedrich replied:
@sandr01d Good catch, thanks a lot. I kept the variable name but changed the implementation to using `tail` instead of `head` and checking against `installed_fzf_version`. Makes the code easier to read IMO. Works just like before. Just to be sure, can you check again? I want to make sure to get this right, as people might be upset when the check does not work correctly.

sandr01d replied:
Works for me, thanks!

sandr01d reviewed (changes requested):
Just one minor thing

sandr01d commented:
I prepared restricting fzf to version >= 0.49.0 for the AUR packages. @cjappl can you handle brew?

cjappl commented:
> I prepared restricting fzf to version >= 0.49.0 for the AUR packages. @cjappl can you handle brew?

Yes, but I'm traveling until 1/29. I can handle it then, or someone else can dive in if they want to before then! I'll set a reminder for myself, just ping me if anyone gets there first.

sandr01d reviewed (approved):
All set from my side, thanks for all the work @carlfriedrich!

cjappl commented:
> I prepared restricting fzf to version >= 0.49.0 for the AUR packages. @cjappl can you handle brew?

This is prepared for a PR and ready to go. When this review is submitted I will put it up for review in the homebrew repo. They are usually very prompt in approving and merging

carlfriedrich commented:
@sandr01d @cjappl Thanks for preparing AUR and brew. This will land in forgit 25.02.0, which will be automatically released on 2025-02-01.

cjappl commented:
Put up for review here for brew:

https://github.com/Homebrew/homebrew-core/pull/205866
